Sources/StreamChat/APIClient/ChatRemoteNotificationHandler.swift

```diff
@@ -0,0 +1,199 @@
+//
+// Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+import UserNotifications
+
+public struct MessageNotificationContent {
+    public let message: ChatMessage
+    public let channel: ChatChannel?
+}
+
+public struct ReactionNotificationContent {
+    public let message: ChatMessage
+    public let channel: ChatChannel?
+}
+
+public struct UnknownNotificationContent {
+    public let content: UNNotificationContent
+}
+
+public enum ChatPushNotificationContent {
+    case message(MessageNotificationContent)
+    case reaction(ReactionNotificationContent)
+    case unknown(UnknownNotificationContent)
+}
+
+enum ChatPushNotificationError: Error {
+    case invalidUserInfo(String)
+}
+
+public struct ChatPushNotificationInfo {
+    public let cid: ChannelId?
+    public let messageId: MessageId?
+    public let eventType: EventType?
+    
+    public init(content: UNNotificationContent) throws {
+        guard let payload = content.userInfo["stream"], let dict = payload as? [String: String] else {
+            throw ChatPushNotificationError.invalidUserInfo("missing stream key or not a [string:string] dict")
+        }
+
+        guard let type = dict["type"] else {
+            throw ChatPushNotificationError.invalidUserInfo("missing stream.type key")
+        }
+
+        eventType = EventType(rawValue: type)
+
+        if let cid = dict["cid"] {
+            self.cid = try? ChannelId(cid: cid)
+        } else {
+            cid = nil
+        }
+
+        if EventType.messageNew.rawValue == type, let id = dict["id"] {
+            messageId = MessageId(id)
+        } else {
+            messageId = nil
+        }
+    }
+}
+
+public class ChatRemoteNotificationHandler {
+    var client: ChatClient
+    var content: UNNotificationContent
+    let chatCategoryIdentifiers: Set<String> = ["stream.chat", "MESSAGE_NEW"]
+    let syncCooldown: TimeInterval = 6.0
+    let database: DatabaseContainer
+
+    public init(client: ChatClient, content: UNNotificationContent) {
+        self.client = client
+        self.content = content
+        database = client.databaseContainer
+    }
+
+    private func obtainLastSyncDate(completion: @escaping (Date?) -> Void) {
+        database.backgroundReadOnlyContext.perform {
+            completion(self.database.backgroundReadOnlyContext.currentUser?.lastReceivedEventDate)
+        }
+    }
+
+    private func bumpLastSyncData(lastReceivedEventDate: Date, completion: @escaping () -> Void) {
+        database.write { session in
+            session.currentUser?.lastReceivedEventDate = lastReceivedEventDate
+            completion()
```

Comment:
there's a `completion` parameter of `database.write`, the `completion` should be called there